Internal error; Cannot execute code in channel thread

[12.10 19:13:01] [Server] [ERROR]   [ProtocolLib] INTERNAL ERROR: Cannot execute code in channel thread.
[12.10 19:13:01] [Server]   If this problem hasn't already been reported, please open a ticket
[12.10 19:13:01] [Server]   at https://github.com/dmulloy2/ProtocolLib/issues with the following data:
[12.10 19:13:01] [Server]   Stack Trace:
[12.10 19:13:01] [Server]   java.lang.IllegalArgumentExceptionDuplicate handler name: protocol_lib_decoder
[12.10 19:13:01] [Server]   	at io.netty.channel.DefaultChannelPipeline.checkDuplicateName(DefaultChannelPipeline.java:1055)
[12.10 19:13:01] [Server]   	at io.netty.channel.DefaultChannelPipeline.filterName(DefaultChannelPipeline.java:284)
[12.10 19:13:01] [Server]   	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:247)
[12.10 19:13:01] [Server]   	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:237)
[12.10 19:13:01] [Server]   	at com.comphenix.protocol.injector.netty.PipelineProxy.addBefore(PipelineProxy.java:61)
[12.10 19:13:01] [Server]   	at com.comphenix.protocol.injector.netty.ChannelInjector$3$1.addBefore(ChannelInjector.java:313)
[12.10 19:13:01] [Server]   	at com.comphenix.protocol.injector.netty.ChannelInjector.inject(ChannelInjector.java:295)
[12.10 19:13:01] [Server]   	at com.comphenix.protocol.injector.netty.ChannelInjector.lambda$executeInChannelThread$3(ChannelInjector.java:917)
[12.10 19:13:01] [Server]   	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
[12.10 19:13:01] [Server]   	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
[12.10 19:13:01] [Server]   	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
[12.10 19:13:01] [Server]   	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
[12.10 19:13:01] [Server]   	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
[12.10 19:13:01] [Server]   	at java.base/java.lang.Thread.run(Thread.java:834)
[12.10 19:13:01] [Server]   Dump:
[12.10 19:13:01] [Server]     manager:
[12.10 19:13:01] [Server]       com.comphenix.protocol.injector.PacketFilterManager@5158ca50[
[12.10 19:13:01] [Server]         unhookTask=com.comphenix.protocol.injector.DelayedSingleTask@4e349edf
[12.10 19:13:01] [Server]         packetListeners=[PacketAdapter[plugin=Plank-Digital, sending=EMPTY_WHITELIST, receiving=ListeningWhitelist[priority=MONITOR, packets=[UPDATE_SIGN[class=PacketPlayInUpdateSign, id=43]], gamephase=PLAYING, options=[]]]]
[12.10 19:13:01] [Server]         packetInjector=com.comphenix.protocol.injector.netty.ProtocolInjector$5@16affb66
[12.10 19:13:01] [Server]         playerInjection=com.comphenix.protocol.injector.netty.ProtocolInjector$4@1c3405b7
[12.10 19:13:01] [Server]         inputBufferedPackets=[]
[12.10 19:13:01] [Server]         recievedListeners=com.comphenix.protocol.injector.SortedPacketListenerList@ba9562b
[12.10 19:13:01] [Server]         sendingListeners=com.comphenix.protocol.injector.SortedPacketListenerList@6181f3e7
[12.10 19:13:01] [Server]         hasClosed=false
[12.10 19:13:01] [Server]         classLoader=PluginClassLoader{plugin=ProtocolLib v4.7.1-SNAPSHOT-b533, pluginEnabled=true, url=plugins/ProtocolLib.jar}
[12.10 19:13:01] [Server]         reporter=com.comphenix.protocol.ProtocolLib$1@89c507a
[12.10 19:13:01] [Server]         server=CraftServer{serverName=Paper,serverVersion=git-Paper-788,minecraftVersion=1.16.5}
[12.10 19:13:01] [Server]         library=ProtocolLib v4.7.1-SNAPSHOT-b533
[12.10 19:13:01] [Server]         asyncFilterManager=com.comphenix.protocol.async.AsyncFilterManager@34b9e563
[12.10 19:13:01] [Server]         knowsServerPackets=true
[12.10 19:13:01] [Server]         knowsClientPackets=true
[12.10 19:13:01] [Server]         phaseLoginCount=0
[12.10 19:13:01] [Server]         phasePlayingCount=1
[12.10 19:13:01] [Server]         packetCreation=false
[12.10 19:13:01] [Server]         nettyInjector=com.comphenix.protocol.injector.netty.ProtocolInjector@a8cd36b
[12.10 19:13:01] [Server]         pluginVerifier=com.comphenix.protocol.injector.PluginVerifier@78b07186
[12.10 19:13:01] [Server]         hasRecycleDistance=true
[12.10 19:13:01] [Server]         minecraftVersion=(MC1.16.5)
[12.10 19:13:01] [Server]         debug=false
[12.10 19:13:01] [Server]       ]
[12.10 19:13:01] [Server]   Sender:
[12.10 19:13:01] [Server]     com.comphenix.protocol.injector.netty.ChannelInjector@62482af7[
[12.10 19:13:01] [Server]       decodeBuffer=protected void net.minecraft.server.v1_16_R3.PacketDecoder.decode(io.netty.channel.ChannelHandlerContext,io.netty.buffer.ByteBuf,java.util.List) throws java.lang.Exception
[12.10 19:13:01] [Server]       encodeBuffer=protected void net.minecraft.server.v1_16_R3.PacketEncoder.encode(io.netty.channel.ChannelHandlerContext,java.lang.Object,io.netty.buffer.ByteBuf) throws java.lang.Exception
[12.10 19:13:01] [Server]       factory=com.comphenix.protocol.injector.netty.InjectionFactory@3ee7af1d
[12.10 19:13:01] [Server]       player=CraftPlayer{name=LauraRussels}
[12.10 19:13:01] [Server]       updated=<null>
[12.10 19:13:01] [Server]       playerName=<null>
[12.10 19:13:01] [Server]       playerConnection=<null>
[12.10 19:13:01] [Server]       networkManager=net.minecraft.server.v1_16_R3.NetworkManager@62d77982
[12.10 19:13:01] [Server]       originalChannel=com.comphenix.protocol.injector.netty.ChannelInjector$3@1130c095
[12.10 19:13:01] [Server]       channelField=VolatileField [accessor=DefaultFieldAccessor [field=public io.netty.channel.Channel net.minecraft.server.v1_16_R3.NetworkManager.channel], container=net.minecraft.server.v1_16_R3.NetworkManager@62d77982, previous=null, current=null, previousLoaded=false, currentSet=false, forceAccess=true]
[12.10 19:13:01] [Server]       packetMarker={}
[12.10 19:13:01] [Server]       currentEvent=<null>
[12.10 19:13:01] [Server]       finalEvent=<null>
[12.10 19:13:01] [Server]       unfilteredProcessedPackets=com.comphenix.protocol.injector.netty.PacketFilterQueue@29db43b5
[12.10 19:13:01] [Server]       vanillaDecoder=net.minecraft.server.v1_16_R3.PacketDecoder@5781becc
[12.10 19:13:01] [Server]       vanillaEncoder=net.minecraft.server.v1_16_R3.PacketEncoder@40a2f8c1
[12.10 19:13:01] [Server]       finishQueue=[]
[12.10 19:13:01] [Server]       channelListener=com.comphenix.protocol.injector.netty.ProtocolInjector@a8cd36b
[12.10 19:13:01] [Server]       processor=com.comphenix.protocol.injector.NetworkProcessor@97b12ed
[12.10 19:13:01] [Server]       injected=false
[12.10 19:13:01] [Server]       closed=false
[12.10 19:13:01] [Server]       cumulation=<null>
[12.10 19:13:01] [Server]       cumulator=io.netty.handler.codec.ByteToMessageDecoder$1@360f4913
[12.10 19:13:01] [Server]       singleDecode=false
[12.10 19:13:01] [Server]       first=false
[12.10 19:13:01] [Server]       firedChannelRead=false
[12.10 19:13:01] [Server]       decodeState=0
[12.10 19:13:01] [Server]       discardAfterReads=16
[12.10 19:13:01] [Server]       numReads=0
[12.10 19:13:01] [Server]       added=true
[12.10 19:13:01] [Server]     ]
[12.10 19:13:01] [Server]   Version:
[12.10 19:13:01] [Server]     ProtocolLib v4.7.1-SNAPSHOT-b533
[12.10 19:13:01] [Server]   Java Version:
[12.10 19:13:01] [Server]     11.0.10
[12.10 19:13:01] [Server]   Server:
[12.10 19:13:01] [Server]     git-Paper-(MC: 1.16.5)

-------


Also side note, is the PacketType.Play.Client.UPDATE_SIGN not supported in 1.16.5 Java 11? When there are no issues with ProtocolLib loading, it throws an error when this packet is received.

Thanks in advance

--- a/src/main/java/com/comphenix/protocol/injector/netty/manager/ListeningList.java
+++ b/src/main/java/com/comphenix/protocol/injector/netty/manager/ListeningList.java
@@ -4,10 +4,14 @@ import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelHandler;
 import java.util.Collection;
+import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
 import java.util.ListIterator;
 import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
 
 @SuppressWarnings("NullableProblems")
 final class ListeningList implements List<Object> {
@@ -25,17 +29,23 @@ final class ListeningList implements List<Object> {
 
 	@Override
 	public int size() {
-		return this.original.size();
+		synchronized (this) {
+			return this.original.size();
+		}
 	}
 
 	@Override
 	public boolean isEmpty() {
-		return this.original.isEmpty();
+		synchronized (this) {
+			return this.original.isEmpty();
+		}
 	}
 
 	@Override
 	public boolean contains(Object o) {
-		return this.original.contains(o);
+		synchronized (this) {
+			return this.original.contains(o);
+		}
 	}
 
 	@Override
@@ -44,80 +54,112 @@ final class ListeningList implements List<Object> {
 	}
 
 	@Override
+	public void forEach(Consumer<? super Object> action) {
+		synchronized (this) {
+			this.original.forEach(action);
+		}
+	}
+
+	@Override
 	public Object[] toArray() {
-		return this.original.toArray();
+		synchronized (this) {
+			return this.original.toArray();
+		}
 	}
 
 	@Override
 	public <T> T[] toArray(T[] a) {
-		//noinspection SuspiciousToArrayCall
-		return this.original.toArray(a);
+		synchronized (this) {
+			//noinspection SuspiciousToArrayCall
+			return this.original.toArray(a);
+		}
 	}
 
 	@Override
 	public boolean add(Object o) {
-		this.processInsert(o);
-		return this.original.add(o);
+		synchronized (this) {
+			this.processInsert(o);
+			return this.original.add(o);
+		}
 	}
 
 	@Override
 	public void add(int index, Object element) {
-		this.processInsert(element);
-		this.original.add(index, element);
+		synchronized (this) {
+			this.processInsert(element);
+			this.original.add(index, element);
+		}
 	}
 
 	@Override
 	public boolean addAll(Collection<?> c) {
-		c.forEach(this::processInsert);
-		return this.original.addAll(c);
+		synchronized (this) {
+			c.forEach(this::processInsert);
+			return this.original.addAll(c);
+		}
 	}
 
 	@Override
 	public boolean addAll(int index, Collection<?> c) {
-		c.forEach(this::processInsert);
-		return this.original.addAll(index, c);
+		synchronized (this) {
+			c.forEach(this::processInsert);
+			return this.original.addAll(index, c);
+		}
 	}
 
 	@Override
 	public Object set(int index, Object element) {
-		this.processInsert(element);
+		synchronized (this) {
+			this.processInsert(element);
 
-		Object prev = this.original.set(index, element);
-		this.processRemove(prev);
+			Object prev = this.original.set(index, element);
+			this.processRemove(prev);
 
-		return prev;
+			return prev;
+		}
 	}
 
 	@Override
 	public boolean remove(Object o) {
-		if (this.original.remove(o)) {
-			this.processRemove(o);
-			return true;
-		}
+		synchronized (this) {
+			if (this.original.remove(o)) {
+				this.processRemove(o);
+				return true;
+			}
 
-		return false;
+			return false;
+		}
 	}
 
 	@Override
 	public boolean containsAll(Collection<?> c) {
-		return this.original.containsAll(c);
+		synchronized (this) {
+			//noinspection SlowListContainsAll
+			return this.original.containsAll(c);
+		}
 	}
 
 	@Override
 	public Object remove(int index) {
-		Object removed = this.original.remove(index);
-		this.processRemove(removed);
-		return removed;
+		synchronized (this) {
+			Object removed = this.original.remove(index);
+			this.processRemove(removed);
+			return removed;
+		}
 	}
 
 	@Override
 	public int indexOf(Object o) {
-		return this.original.indexOf(o);
+		synchronized (this) {
+			return this.original.indexOf(o);
+		}
 	}
 
 	@Override
 	public int lastIndexOf(Object o) {
-		return this.original.lastIndexOf(o);
+		synchronized (this) {
+			return this.original.lastIndexOf(o);
+		}
 	}
 
 	@Override
@@ -132,33 +174,77 @@ final class ListeningList implements List<Object> {
 
 	@Override
 	public List<Object> subList(int fromIndex, int toIndex) {
-		return this.original.subList(fromIndex, toIndex);
+		synchronized (this) {
+			return this.original.subList(fromIndex, toIndex);
+		}
 	}
 
 	@Override
 	public boolean removeAll(Collection<?> c) {
-		c.forEach(element -> {
-			if (this.original.contains(element)) {
-				this.processRemove(element);
-			}
-		});
-		return this.original.removeAll(c);
+		synchronized (this) {
+			c.forEach(element -> {
+				if (this.original.contains(element)) {
+					this.processRemove(element);
+				}
+			});
+			return this.original.removeAll(c);
+		}
+	}
+
+	@Override
+	public boolean removeIf(Predicate<? super Object> filter) {
+		synchronized (this) {
+			return this.original.removeIf(object -> {
+				boolean shouldRemove = filter.test(object);
+				if (shouldRemove) {
+					this.processRemove(object);
+				}
+				return shouldRemove;
+			});
+		}
 	}
 
 	@Override
 	public boolean retainAll(Collection<?> c) {
-		return this.original.retainAll(c);
+		synchronized (this) {
+			return this.original.retainAll(c);
+		}
+	}
+
+	@Override
+	public void replaceAll(UnaryOperator<Object> operator) {
+		synchronized (this) {
+			this.original.replaceAll(value -> {
+				Object newValue = operator.apply(value);
+				if (newValue != value) {
+					this.processRemove(value);
+					this.processInsert(newValue);
+				}
+				return newValue;
+			});
+		}
+	}
+
+	@Override
+	public void sort(Comparator<? super Object> c) {
+		synchronized (this) {
+			this.original.sort(c);
+		}
 	}
 
 	@Override
 	public void clear() {
-		this.original.forEach(this::processRemove);
-		this.original.clear();
+		synchronized (this) {
+			this.original.forEach(this::processRemove);
+			this.original.clear();
+		}
 	}
 
 	@Override
 	public Object get(int index) {
-		return this.original.get(index);
+		synchronized (this) {
+			return this.original.get(index);
+		}
 	}
 
 	private void processInsert(Object element) {

--- a/src/main/java/com/comphenix/protocol/injector/netty/manager/NetworkManagerInjector.java
+++ b/src/main/java/com/comphenix/protocol/injector/netty/manager/NetworkManagerInjector.java
@@ -26,7 +26,6 @@ import io.netty.channel.ChannelFuture;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
-import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
@@ -205,7 +204,7 @@ public class NetworkManagerInjector implements ChannelListener {
 				//noinspection SynchronizationOnLocalVariableOrMethodParameter
 				synchronized (value) {
 					// override the list field with our list
-					List<Object> newList = Collections.synchronizedList(new ListeningList(value, this.pipelineInjectorHandler));
+					List<Object> newList = new ListeningList(value, this.pipelineInjectorHandler);
 					accessor.set(serverConnection, newList);
 				}
 			}
@@ -215,7 +214,6 @@ public class NetworkManagerInjector implements ChannelListener {
 		this.injected = true;
 	}
 
-	@SuppressWarnings("unchecked")
 	public void close() {
 		if (this.closed || !this.injected) {
 			return;
@@ -229,10 +227,11 @@ public class NetworkManagerInjector implements ChannelListener {
 		for (Pair<Object, FieldAccessor> list : this.overriddenLists) {
 			// get the value of the field we've overridden, if it is no longer a ListeningList someone probably jumped in
 			// and replaced the field himself - we are out safely as the other person needs to clean the mess...
-			List<Object> value = (List<Object>) list.getSecond().get(list.getFirst());
-			if (value instanceof ListeningList) {
+			// just reset to the list we wrapped originally
+			Object currentFieldValue = list.getSecond().get(list.getFirst());
+			if (currentFieldValue instanceof ListeningList) {
 				// just reset to the list we wrapped originally
-				ListeningList ourList = (ListeningList) value;
+				ListeningList ourList = (ListeningList) currentFieldValue;
 				List<Object> original = ourList.getOriginal();
 				//noinspection SynchronizationOnLocalVariableOrMethodParameter
 				synchronized (original) {

--- a/src/main/java/com/comphenix/protocol/injector/netty/manager/NetworkManagerInjector.java
+++ b/src/main/java/com/comphenix/protocol/injector/netty/manager/NetworkManagerInjector.java
@@ -227,7 +227,6 @@ public class NetworkManagerInjector implements ChannelListener {
 		for (Pair<Object, FieldAccessor> list : this.overriddenLists) {
 			// get the value of the field we've overridden, if it is no longer a ListeningList someone probably jumped in
 			// and replaced the field himself - we are out safely as the other person needs to clean the mess...
-			// just reset to the list we wrapped originally
 			Object currentFieldValue = list.getSecond().get(list.getFirst());
 			if (currentFieldValue instanceof ListeningList) {
 				// just reset to the list we wrapped originally

